src/transformers/models/pop2piano/tokenization_pop2piano.py

```diff
@@ -0,0 +1,352 @@
+        for type, number in words:
+            if type == self.encoder["TOKEN_SPECIAL"]:
+                if number == 1:
+                    break
+            elif type == self.encoder["TOKEN_TIME"]:
+                current_idx += number
+                if cutoff_time_idx is not None:
+                    current_idx = min(current_idx, cutoff_time_idx)
+
+            elif type == self.encoder["TOKEN_VELOCITY"]:
+                current_velocity = number
+            elif type == self.encoder["TOKEN_NOTE"]:
+                pitch = number
+                if current_velocity == 0:
+                    # note_offset
+                    if note_onsets_ready[pitch] is None:
+                        # offset without onset
+                        pass
+                    else:
+                        onset_idx = note_onsets_ready[pitch]
+                        if onset_idx >= current_idx:
+                            # No time shift after previous note_on
+                            pass
+                        else:
+                            offset_idx = current_idx
+                            notes.append([onset_idx, offset_idx, pitch, self.encoder["DEFAULT_VELOCITY"]])
+                            note_onsets_ready[pitch] = None
+                else:
+                    # note_on
+                    if note_onsets_ready[pitch] is None:
+                        note_onsets_ready[pitch] = current_idx
+                    else:
+                        # note-on already exists
+                        onset_idx = note_onsets_ready[pitch]
+                        if onset_idx >= current_idx:
+                            # No time shift after previous note_on
+                            pass
+                        else:
+                            offset_idx = current_idx
+                            notes.append([onset_idx, offset_idx, pitch, self.encoder["DEFAULT_VELOCITY"]])
+                            note_onsets_ready[pitch] = current_idx
+            else:
+                raise ValueError("Token type not understood!")
+
+        for pitch, note_on in enumerate(note_onsets_ready):
+            # force offset if no offset for each pitch
+            if note_on is not None:
+                if cutoff_time_idx is None:
+                    cutoff = note_on + 1
+                else:
+                    cutoff = max(cutoff_time_idx, note_on + 1)
+
+                offset_idx = max(current_idx, cutoff)
+                notes.append([note_on, offset_idx, pitch, self.encoder["DEFAULT_VELOCITY"]])
+
+        if len(notes) == 0:
+            return []
+        else:
+            notes = np.array(notes)
+            note_order = notes[:, 0] * 128 + notes[:, 1]
+            notes = notes[note_order.argsort()]
+            return notes
+
+    def notes_to_midi(self, notes, beatstep, offset_sec=None):
+        """Converts notes to midi"""
+
+        new_pm = pretty_midi.PrettyMIDI(resolution=384, initial_tempo=120.0)
```

Comment:
I guess we need a package here as well?

Reply:
yes we must need it for the tokenizer.

Reply:
We could write this ourselves, I've got code for it somewhere. Writing MIDI is pretty simple. On the other hand, we do need to return the result in an object of some kind (if we're not saving to a file) and so we might as well use PrettyMIDI.

Reply:
If you have the PrettyMidi code then please share it here I will change it when I get time. And regarding the return object, isn't returning a PrettyMidi object sufficient? Or should I create a class `PrettyMIDIOutput`?

Reply:
ooh got it! so you mean to pick the best MIDI library so that in future we can stick to it for future models. Well, I don't know :(,  honestly speaking before this integration I didn't even knew what MIDI is...I believe @sweetcocoa can guide us here.

Reply:
Well, As far as I know, [PrettyMIDI](https://github.com/craffel/pretty-midi) is the most popular MIDI library in the MIR field. Of course, there are more libraries other than [PrettyMIDI](https://github.com/craffel/pretty-midi), such as [Mido](https://github.com/mido/mido), which is the core backend of [PrettyMIDI](https://github.com/craffel/pretty-midi), and [note_seq](https://github.com/magenta/note-seq) which is used by Google Magenta. but note_seq is not maintained these days and it uses prettymidi too as its dependency.

Many popular open-source MIR projects are built on top of these libraries, such as 
 - [Google Magenta](https://github.com/magenta/magenta/) uses PrettyMidi and Mido.
 - [Omnizart](https://github.com/Music-and-Culture-Technology-Lab/omnizart) uses PrettyMidi.
 - [Regression Onsets-and-Frames](https://github.com/bytedance/piano_transcription) uses Mido.
 - and other projects that i've not mentioned here.


I think other python midi libraries are not suitable, in the point of generality and popularity. ( Of course, if there are other popular enough libraries that I've missed, I could be wrong. )



So here, Mido and PrettyMidi are worth consider, and i think PrettyMidi is more suitable because it provides many utility functions to modulate midi and more human-readable class.

Reply:
Awesome, thanks @sweetcocoa. 

MIDI is really just a sequence of events such as Note On and Note Off, so if we *really* don't want to pull in another package, we could just return a list of tuples containing those events and let the user figure out what to do with it. But from the sounds of it, PrettyMIDI is pretty much the standard way to manipulate MIDI in Python, and so the user would end up importing the sequence into PrettyMIDI anyway.